src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -293,6 +294,69 @@ fun processMacroSimpleResolveVariants(element: RsMacroBodySimpleMatching, proces
     return simple.any { processor(it) }
 }
 
+fun processMacroReferenceVariants(referenceElement: RsReferenceElement, isCompletion: Boolean, processor: RsResolveProcessor): Boolean {
+    val prevScope = mutableSetOf<String>()
+    walkUp(referenceElement, { false }) { cameFrom, scope ->
+        val currScope = mutableListOf<String>()
+        val shadowingProcessor = { e: ScopeEntry ->
+            e.name !in prevScope && run {
+                currScope += e.name
+                processor(e)
+            }
+        }
+        if (processMacroDeclarations(scope, cameFrom, shadowingProcessor)) return@walkUp true
+        prevScope.addAll(currScope)
+        false
+    }
+    return false
+}
+
+fun processMacroDeclarations(scope: RsCompositeElement, cameFrom: RsCompositeElement, processor: RsResolveProcessor): Boolean {
+    check(cameFrom.context == scope)
+    when (scope) {
+        is RsFile -> if (processItemMacroDeclarations(scope, processor, false, scope.isCrateRoot)) return true
+        is RsMod -> if (processItemMacroDeclarations(scope, processor)) return true
+        is RsBlock -> if (processItemMacroDeclarations(scope, processor)) return true
+    }
+    return false
+}
+
+fun RsExternCrateItem.findOuterAttr(name: String): RsOuterAttr? =
+    outerAttrList.find { it.metaItem.identifier.textMatches(name) }
+
+fun RsInnerAttributeOwner.findInnerAttr(name: String): RsInnerAttr? =
+    innerAttrList.find { it.metaItem.identifier.textMatches(name) }
+
+fun processItemMacroDeclarations(
+    scope: RsItemsOwner,
+    processor: RsResolveProcessor,
+    need_export: Boolean = false,
+    add_std_crate: Boolean = false
+): Boolean {
+    val macros = scope.macroItemList.filter { !need_export || it.hasMacroExport }
+    if (processAll(macros, processor)) return true
+
+    for (crate in scope.externCrateItemList.filter { it.hasMacroUse }) {// .filter { it.findOuterAttr("macro_use") != null }
+        val mod = crate.reference.resolve() ?: continue
+        if (mod is RsMod && processItemMacroDeclarations(mod, processor, true)) return true
+    }
+    if (add_std_crate) {
+        val stdCrate = scope.containingCargoPackage?.findCrateByName("std")?.crateRoot
+        val prelude = scope.project.getPsiFor(stdCrate)?.rustMod
+        if (prelude is RsMod && processItemMacroDeclarations(prelude, processor, true)) return true
+    }
+
+    for (modDecl in scope.modDeclItemList.filter { it.hasMacroUse }) {
+        val mod = modDecl.reference.resolve() ?: continue
+        if (mod is RsMod && processItemMacroDeclarations(mod, processor, true)) return true
```

Comment:
So in the worst case, we'll walk the whole crate here, but we hope only some modules have `macro_use`?

I think it is very sensible approach, though I think that perhaps we need to cache this information? Something like 

```
fun exportedMacros(crateRoot: RsFille): Collection<RsMacroDefinition> {
     CachedValuesManager.getCachedValue ... 
}
```